File: year2019/day04/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input: number[] = file.toString().split('-').map(Number);
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return Array.from({ length: input[1] - input[0] + 1 }, (_, index) => (input[0] + index).toString()).filter(pw => /(.)\1/.test(pw) && pw === pw.split('').sort().join('')).length;
}

export function part_2(): number {
  return Array.from({ length: input[1] - input[0] + 1 }, (_, index) => (input[0] + index).toString()).filter(pw => pw === pw.split('').sort().join('') && pw.split('').sort().some((x, i, a) => x === a[i + 1] && x !== a[i + 2] && x !== a[i - 1])).length;
}

File: year2019/day04/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return len([pw for pw in map(str, range(input[0], input[1] + 1)) if len(set(pw)) < len(pw) and pw == ''.join(sorted(pw))])


def part_2():
  return len([pw for pw in map(str, range(input[0], input[1] + 1)) if pw == ''.join(sorted(pw)) and any(pw[i] == pw[i + 1] and (i == 0 or pw[i] != pw[i - 1]) and (i + 2 == len(pw) or pw[i] != pw[i + 2]) for i in range(len(pw) - 1))])


def setup(path):
  global input

  with open(path, 'r') as file:
    input = tuple(map(int, file.read().split('-')))


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()